File: PokemonGame_CSharp/Battle.cs
public class Battle
{
    public Trainer Trainer1 { get; set; }
    public Trainer Trainer2 { get; set; }

    public Battle(Trainer trainer1, Trainer trainer2)
    {
        Trainer1 = trainer1;
        Trainer2 = trainer2;
    }

    public string FightRound()
    {
        Trainer1.ThrowPokeball();
        Trainer2.ThrowPokeball();

        var pokemon1 = Trainer1.ActivePokeball.Pokemon;
        var pokemon2 = Trainer2.ActivePokeball.Pokemon;

        if (pokemon1.Strength == pokemon2.Weakness)
        {
            Trainer2.ReturnPokemon();
            Trainer1.IncreaseCount();
            return $"{Trainer1.Name}'s {pokemon1.Nickname} wins the round!";
        }
        else if (pokemon2.Strength == pokemon1.Weakness)
        {
            Trainer1.ReturnPokemon();
            Trainer2.IncreaseCount();
            return $"{Trainer2.Name}'s {pokemon2.Nickname} wins the round!";
        }
        else
        {
            // If neither Pokemon has a strength that is the other's weakness, decide the winner based on a predefined type advantage order
            if ((pokemon1.Strength == ElementType.Fire && pokemon2.Strength == ElementType.Grass) ||
                (pokemon1.Strength == ElementType.Grass && pokemon2.Strength == ElementType.Water) ||
                (pokemon1.Strength == ElementType.Water && pokemon2.Strength == ElementType.Fire))
            {
                Trainer2.ReturnPokemon();
                return $"{Trainer1.Name}'s {pokemon1.Nickname} wins the round!";
            }
            else
            {
                Trainer1.ReturnPokemon();
                return $"{Trainer2.Name}'s {pokemon2.Nickname} wins the round!";
            }
        }
    }

  public string FightBattle()
{
    int rounds = 0;
    try
    {
        while (Trainer1.Count > 0 && Trainer2.Count > 0)
        {
            Console.WriteLine(FightRound());
            rounds++;
        }
    }
    catch (Exception ex)
    {
        Console.WriteLine(ex.Message);
    }

    if (Trainer1.Count > 0 && Trainer2.Count > 0)
    {
        return $"The battle is a draw after {rounds} rounds!";
    }
    else if (Trainer1.Count == 0)
    {
        return $"{Trainer2.Name} wins the battle after {rounds} rounds!";
    }
    else
    {
        return $"{Trainer1.Name} wins the battle after {rounds} rounds!";
    }
}
}

File: Pokemon_Game/Battle.py
import random

class Battle:
    def __init__(self, trainer1, trainer2):
        self.trainer1 = trainer1
        self.trainer2 = trainer2
        self.winner_of_last_round = None

    def conduct_round(self):
        from Arena import Arena  # Move the import statement here

        pokeball1 = random.choice(self.trainer1.belt)
        pokeball2 = random.choice(self.trainer2.belt)
        pokemon1 = pokeball1.throw()
        pokemon2 = pokeball2.throw()

        print(f"{self.trainer1.name} throws a pokeball!")
        print(f"{pokemon1.nickname} says: {pokemon1.battle_cry()}")
        print(f"{self.trainer2.name} throws a pokeball!")
        print(f"{pokemon2.nickname} says: {pokemon2.battle_cry()}")

        if pokemon1.strength == pokemon2.weakness:
            self.winner_of_last_round = self.trainer1
            self.trainer2.belt.remove(pokeball2)
            print(f"{pokemon1.nickname} wins the round because {pokemon1.nickname}'s strength ({pokemon1.strength}) is {pokemon2.nickname}'s weakness ({pokemon2.weakness}). {pokemon2.nickname} is defeated!")
        elif pokemon2.strength == pokemon1.weakness:
            self.winner_of_last_round = self.trainer2
            self.trainer1.belt.remove(pokeball1)
            print(f"{pokemon2.nickname} wins the round because {pokemon2.nickname}'s strength ({pokemon2.strength}) is {pokemon1.nickname}'s weakness ({pokemon1.weakness}). {pokemon1.nickname} is defeated!")
        else:
            if self.winner_of_last_round == self.trainer1:
                self.trainer2.belt.remove(pokeball2)
                print(f"{pokemon1.nickname} wins the round because it won the previous round. {pokemon2.nickname} is defeated!")
            elif self.winner_of_last_round == self.trainer2:
                self.trainer1.belt.remove(pokeball1)
                print(f"{pokemon2.nickname} wins the round because it won the previous round. {pokemon1.nickname} is defeated!")
            else:
                self.trainer1.belt.remove(pokeball1)
                self.trainer2.belt.remove(pokeball2)
                print("It's a draw! Both Pokemon are returned to their Pokeballs.")

        print(f"{self.trainer1.name} has {len(self.trainer1.belt)} Pokeballs left.")
        print(f"{self.trainer2.name} has {len(self.trainer2.belt)} Pokeballs left.")

        # Increment the total rounds after each round
        Arena.increment_rounds()

    def start_battle(self):
        while len(self.trainer1.belt) > 0 and len(self.trainer2.belt) > 0:
            self.conduct_round()

    def get_winner(self):
        if len(self.trainer1.belt) > len(self.trainer2.belt):
            return self.trainer1
        elif len(self.trainer2.belt) > len(self.trainer1.belt):
            return self.trainer2
        else:
            return None